syntax = "proto3";

package Serialisation;

message Coordinates {
    double lat = 1;
    double lng = 2;
}

message Stop {
    string stop_name = 1;
    Coordinates stop_coords = 2;
}

message Rgb {
    int32 red = 1;
    int32 green = 2;
    int32 blue = 3;
}

message Rgba {
    int32 red = 1;
    int32 green = 2;
    int32 blue = 3;
    double opacity = 4;
}

message Color {
    string string_color = 1;
    Rgb rgb_color = 2;
    Rgba rgba_color = 3;
}

message Bus {
    Color color = 1;
    string bus_name = 2;
    repeated string bus_stops = 3;
    int32 unique_stops_count = 4;
    double route_unreal_lenght = 5;
    double route_real_lenght = 6;
}

message render_settings {
    double width = 1;
    double height = 2;
    double padding = 3;
    double line_width = 4;
    double stop_radius = 5;
    int32 bus_label_font_size = 6;
    double bus_label_offset_1 = 7;
    double bus_label_offset_2 = 8;
    int32 stop_label_font_size = 9;
    double stop_label_offset_1 = 10;
    double stop_label_offset_2 = 11;
    Color underlayer_color = 12;
    double underlayer_width = 13;
    repeated Color color_palette = 14;
}

message routing_settings {
    int32 bus_wait_time = 1;
    double bus_velocity = 2;
}

message AditionalInfo {
    string bus_name = 1;
    string last_stop = 2;
}

message GraphEdge {
    uint32 from = 1;
    uint32 to_ = 2;
    double weight = 3;
    string bus_name = 4;
    int32 span_count = 5;  
}

message StopNameToVertexId {
    string stop_name = 1;
    uint32 id = 2;
}

message DataBase {
    repeated Bus buses = 1;
    repeated Stop stops = 2;
    render_settings render_sett = 3;
    routing_settings routing_sett = 4;
    repeated AditionalInfo aditional_info = 5;
    repeated GraphEdge graph = 6;
    repeated StopNameToVertexId stopname_to_vertex_id = 7;
}